_posts/2018-11-07-retired-pitfalls-22.md

```diff
@@ -0,0 +1,153 @@
+---
+layout: post
+title: Retiring old performance pitfalls
+tags: erts compiler
+author: John Högberg
+---
+
+Erlang/OTP 22 brings many performance improvements to the table, but most of
+them have a broad impact and don't affect the way you write efficient code. In
+this post I'd like to highlight a few things that used to be surprisingly slow
+but no longer need to be avoided.
+
+### Named fun recursion
+
+Named funs have a neat little feature that might not be obvious at a first
+glance; their name is a variable like any other and you're free to pass it to
+another function or even return it.
+
+This is cool but a bit of a headache for the compiler. To create a fun we pass
+its definition and free variables to a `make_fun2` instruction, but we can't
+include the fun itself as a free variable because it hasn't been created yet.
+Prior to OTP 22 we solved this by creating a new equivalent fun _inside_ the
+fun, but this made recursion surprisingly expensive both in terms of run-time
+and memory use.
+
+As of OTP 22 we translate recursion to a direct function call instead which
+avoids creating a new fun. Other cases still require recreating the fun, but
+they're far less common.
+
+[Optimize named funs and fun-wrapped macros #1973](https://github.com/erlang/otp/pull/1973)
+
+### List subtraction with large operands (-- operator)
+
+While the Erlang VM appears to be pre-emptively scheduled to the programmer,
+it's co-operatively scheduled internally. When a native function runs it
+monopolizes the scheduler until it returns so a long-running one can severely
+harm the responsiveness of the system. We've therefore written nearly all such
+functions in a style that breaks the work into short units that complete
+quickly enough, but there's a steadily shrinking list of functions that
+misbehave, and list subtraction was one of these.
+
+It's usually pretty straightforward to rewrite functions in this style, but
+the old algorithm processed the second list in a loop around the first list,
+which is problematic since both lists can be very long and resuming work in
+nested loops is almost always trickier than expected.
+
+In this case it was easier to just get rid of the nested loop altogether. The
```

Comment:
Maybe add links to the source code? So that the reader can follow in there if she wants.